### DISCUSSION TITLE
Check if retained message has the same contents don't resend it

### DISCUSSION THREAD

**Discussion by skinkie** [halfgaar/FlashMQ#72] [General]
Is it possible, on server level, to prevent retained message with the exact same contents to be resend?

So something like:
```
if incomming.message == retained messages [incomming.topic] then
   noop
```

**Comment by halfgaar** [comment]
I'm not sure what you mean. A client gets retained messages on subscription, so you when you subscribe to a path, under what condition do you not want to receive the message?

**Reply by skinkie** [reply]
If the message that is updated by the publisher, is equal to the current retained message. Considering some proces that publishes a temperature, at a specific rate. The receiving client is only interested in this data if it changes. This simple case could be handled by the publisher it likely knows the previous value. In a multi publisher situation, retaining the last state locally does not make sense. So I wonder if the server could be more intelligent than just forwarding _any_ message.

**Reply by halfgaar** [reply]
At least when it comes to the spec, your wishes are reversed. What the spec specifies, is that a message always just follows normal publish-to-subscriber paths, and when the 'retain' bit it set, it also sets the value as retained.

You could filter this out in the plugin, but only in a limited fashion. You can keep track of (the hashes of) all the messages you've seen and reject it if you've seen it, but that won't help you on a fresh start with an existing database of retained messages.

Your temperature example has problems that I dealt with in a real-life environment. If you only publish when it changes, new subscribers won't see it. At least, they will only see the retained value, and that may have been set a week ago. Retained messages make more sense for something like a sensor ID, or perhaps some settings it has, that are only published on change.
